src/StructuredLogger/ObjectModel/TreeNode.cs

```diff
@@ -47,6 +47,8 @@ public bool IsExpanded
             }
         }
 
+        private Dictionary<ChildrenCacheKey, object> childrenCache;

```

Comment:
My only concern is that every object in the tree will become 8 bytes larger (the viewer is a 64-bit app) due to this new field. I've worked very hard to fine tune the object layout and minimize fields on base classes such as TreeNode.

How about instead of storing the childrenCache in a separate field you make a class ChildrenList : List<object> which will also store the dictionary and the caching logic. Then you can just switch to using your custom collection as storage for Children, and in the Seal() method when you ToArray() that, it just goes away and the unnecessary dictionary will get collected.

What do you think about that?